File: app/src/main/java/net/j4dy/familypicingame/games/whack/WhackGameState.kt
package net.j4dy.familypicingame.games.whack

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.Offset
import net.j4dy.familypicingame.model.FaceProfile
import kotlin.random.Random

enum class PortalType {
    EMPTY,
    MONSTER,
    FAMILY
}

class PortalState(
    val index: Int
) {
    var type by mutableStateOf(PortalType.EMPTY)
    var familyProfile by mutableStateOf<FaceProfile?>(null)
    var showTimeLeft by mutableStateOf(0) // frames it stays visible
}

class WhackSparkle(
    val index: Int,
    val offset: Offset,
    val text: String,
    val isPenalty: Boolean,
    var alpha: Float = 1.0f
)

class WhackGameState(
    val allProfiles: List<FaceProfile>
) {
    val portals = List(9) { PortalState(it) }
    val sparkles = mutableStateListOf<WhackSparkle>()
    
    var score by mutableStateOf(0)
    var comboMultiplier by mutableStateOf(1)
    var timeLeftSeconds by mutableStateOf(45)
    
    var isPlaying by mutableStateOf(false)
    var isGameOver by mutableStateOf(false)
    var gameMessage by mutableStateOf("TAP START TO WHACK")
    
    private val defaultRoundTime = 45
    private var framesSinceLastSpawn = 0
    
    // Scale speed as score gets higher
    private fun getSpawnInterval(): Int {
        return when {
            score > 5000 -> 30 // spawn every 30 frames (~500ms)
            score > 3000 -> 40
            score > 1000 -> 50
            else -> 60 // ~1s
        }
    }
    
    private fun getPortalDuration(): Int {
        return when {
            score > 5000 -> 45 // stays up for 750ms
            score > 3000 -> 60
            score > 1000 -> 75
            else -> 90 // stays up for 1.5s
        }
    }

    fun startGame() {
        isPlaying = true
        isGameOver = false
        score = 0
        comboMultiplier = 1
        timeLeftSeconds = defaultRoundTime
        sparkles.clear()
        framesSinceLastSpawn = 0
        portals.forEach {
            it.type = PortalType.EMPTY
            it.familyProfile = null
            it.showTimeLeft = 0
        }
        gameMessage = "Round Started! Go!"
    }

    fun resetGame() {
        isPlaying = false
        isGameOver = false
        score = 0
        comboMultiplier = 1
        timeLeftSeconds = defaultRoundTime
        sparkles.clear()
        portals.forEach {
            it.type = PortalType.EMPTY
            it.familyProfile = null
            it.showTimeLeft = 0
        }
        gameMessage = "TAP START TO WHACK"
    }

    fun tickSecond() {
        if (!isPlaying || isGameOver) return
        timeLeftSeconds--
        if (timeLeftSeconds <= 0) {
            endGame()
        }
    }

    fun tickFrame() {
        if (!isPlaying || isGameOver) return
        
        // 1. Tick existing portal durations
        portals.forEach { portal ->
            if (portal.type != PortalType.EMPTY) {
                portal.showTimeLeft--
                if (portal.showTimeLeft <= 0) {
                    // Reset to empty when time expires
                    portal.type = PortalType.EMPTY
                    portal.familyProfile = null
                }
            }
        }
        
        // 2. Spawn new characters randomly
        framesSinceLastSpawn++
        if (framesSinceLastSpawn >= getSpawnInterval()) {
            framesSinceLastSpawn = 0
            spawnRandomPortal()
        }
        
        // 3. Tick sparkles alpha decay
        val iterator = sparkles.iterator()
        while (iterator.hasNext()) {
            val sparkle = iterator.next()
            sparkle.alpha -= 0.05f
        }
        sparkles.removeAll { it.alpha <= 0f }
    }

    private fun spawnRandomPortal() {
        // Find empty portals
        val emptyPortals = portals.filter { it.type == PortalType.EMPTY }
        if (emptyPortals.isEmpty()) return
        
        // Select a random empty portal
        val portal = emptyPortals[Random.nextInt(emptyPortals.size)]
        
        // 70% chance of Monster, 30% chance of Family member
        val isFamily = Random.nextFloat() < 0.30f && allProfiles.isNotEmpty()
        
        if (isFamily) {
            portal.type = PortalType.FAMILY
            // Pick a random family member
            portal.familyProfile = allProfiles[Random.nextInt(allProfiles.size)]
        } else {
            portal.type = PortalType.MONSTER
            portal.familyProfile = null
        }
        
        portal.showTimeLeft = getPortalDuration()
    }

    fun whackCell(index: Int) {
        if (!isPlaying || isGameOver || index !in 0..8) return
        
        val portal = portals[index]
        when (portal.type) {
            PortalType.MONSTER -> {
                val points = 100 * comboMultiplier
                score += points
                comboMultiplier++
                
                // Add positive feedback sparkle
                sparkles.add(
                    WhackSparkle(
                        index = index,
                        offset = Offset(Random.nextFloat() * 40 - 20, -50f),
                        text = "+$points",
                        isPenalty = false
                    )
                )
                
                gameMessage = "HIT! Combo x$comboMultiplier"
                portal.type = PortalType.EMPTY
                portal.familyProfile = null
            }
            PortalType.FAMILY -> {
                // Tapping family member is a penalty
                val penalty = 200
                score = (score - penalty).coerceAtLeast(0)
                comboMultiplier = 1
                
                val familyName = portal.familyProfile?.name ?: "Family"
                // Add penalty feedback sparkle
                sparkles.add(
                    WhackSparkle(
                        index = index,
                        offset = Offset(Random.nextFloat() * 40 - 20, -50f),
                        text = "Don't tap $familyName! -$penalty",
                        isPenalty = true
                    )
                )
                
                gameMessage = "OW! That's $familyName!"
                portal.type = PortalType.EMPTY
                portal.familyProfile = null
            }
            PortalType.EMPTY -> {
                // Optional minor miss penalty
                comboMultiplier = 1
                gameMessage = "MISS!"
            }
        }
    }

    private fun endGame() {
        isGameOver = true
        isPlaying = false
        gameMessage = "TIME'S UP! Game Over!"
    }
}

File: lib/games/whack/whack_game_state.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import '../../models/face_profile.dart';

enum PortalType { empty, teamA, teamB }

class PortalState {
  final int index;
  PortalType type;
  FaceProfile? familyProfile;
  int showTimeLeft; // frames it stays visible

  PortalState({
    required this.index,
    this.type = PortalType.empty,
    this.familyProfile,
    this.showTimeLeft = 0,
  });
}

class WhackSparkle {
  final int index;
  final Offset offset;
  final String text;
  final bool isPenalty;
  double alpha;

  WhackSparkle({
    required this.index,
    required this.offset,
    required this.text,
    required this.isPenalty,
    this.alpha = 1.0,
  });
}

class WhackGameState {
  final List<FaceProfile> teamAProfiles;
  final List<FaceProfile> teamBProfiles;
  final double speedMultiplier;

  final List<PortalState> portals = List.generate(12, (i) => PortalState(index: i));
  final List<WhackSparkle> sparkles = [];

  int score = 0;
  int comboMultiplier = 1;
  int timeLeftSeconds = 45;

  bool isPlaying = false;
  bool isGameOver = false;
  String gameMessage = "TAP START TO PLAY";

  final int defaultRoundTime = 45;
  int _framesSinceLastSpawn = 0;

  final _random = math.Random();

  WhackGameState({
    required this.teamAProfiles,
    required this.teamBProfiles,
    this.speedMultiplier = 2.0,
  });

  int _getSpawnInterval() {
    int baseInterval;
    if (score > 5000) {
      baseInterval = 30; // ~500ms
    } else if (score > 3000) {
      baseInterval = 40;
    } else if (score > 1000) {
      baseInterval = 50;
    } else {
      baseInterval = 60; // ~1s
    }
    return (baseInterval / speedMultiplier).round().clamp(10, 100);
  }

  int _getPortalDuration() {
    int baseDuration;
    if (score > 5000) {
      baseDuration = 45; // ~750ms
    } else if (score > 3000) {
      baseDuration = 60;
    } else if (score > 1000) {
      baseDuration = 75;
    } else {
      baseDuration = 90; // ~1.5s
    }
    return (baseDuration / speedMultiplier).round().clamp(15, 120);
  }

  void startGame() {
    isPlaying = true;
    isGameOver = false;
    score = 0;
    comboMultiplier = 1;
    timeLeftSeconds = defaultRoundTime;
    sparkles.clear();
    _framesSinceLastSpawn = 0;
    for (final p in portals) {
      p.type = PortalType.empty;
      p.familyProfile = null;
      p.showTimeLeft = 0;
    }
    gameMessage = "Go Team! Speed: ${speedMultiplier}x!";
  }

  void resetGame() {
    isPlaying = false;
    isGameOver = false;
    score = 0;
    comboMultiplier = 1;
    timeLeftSeconds = defaultRoundTime;
    sparkles.clear();
    for (final p in portals) {
      p.type = PortalType.empty;
      p.familyProfile = null;
      p.showTimeLeft = 0;
    }
    gameMessage = "TAP START TO PLAY";
  }

  void tickSecond() {
    if (!isPlaying || isGameOver) return;
    timeLeftSeconds--;
    if (timeLeftSeconds <= 0) {
      _endGame();
    }
  }

  void tickFrame() {
    if (!isPlaying || isGameOver) return;

    // 1. Tick existing portal durations
    for (final portal in portals) {
      if (portal.type != PortalType.empty) {
        portal.showTimeLeft--;
        if (portal.showTimeLeft <= 0) {
          portal.type = PortalType.empty;
          portal.familyProfile = null;
        }
      }
    }

    // 2. Spawn new characters
    _framesSinceLastSpawn++;
    if (_framesSinceLastSpawn >= _getSpawnInterval()) {
      _framesSinceLastSpawn = 0;
      _spawnRandomPortal();
    }

    // 3. Tick sparkles alpha decay
    for (int i = sparkles.length - 1; i >= 0; i--) {
      sparkles[i].alpha -= 0.05;
      if (sparkles[i].alpha <= 0.0) {
        sparkles.removeAt(i);
      }
    }
  }

  void _spawnRandomPortal() {
    final emptyPortals = portals.where((p) => p.type == PortalType.empty).toList();
    if (emptyPortals.isEmpty) return;

    final portal = emptyPortals[_random.nextInt(emptyPortals.length)];
    final isTeamA = _random.nextDouble() < 0.50;

    if (isTeamA && teamAProfiles.isNotEmpty) {
      portal.type = PortalType.teamA;
      portal.familyProfile = teamAProfiles[_random.nextInt(teamAProfiles.length)];
    } else if (teamBProfiles.isNotEmpty) {
      portal.type = PortalType.teamB;
      portal.familyProfile = teamBProfiles[_random.nextInt(teamBProfiles.length)];
    } else {
      portal.type = PortalType.empty;
      portal.familyProfile = null;
    }

    if (portal.type != PortalType.empty) {
      portal.showTimeLeft = _getPortalDuration();
    }
  }

  void whackCell(int index) {
    if (!isPlaying || isGameOver || index < 0 || index >= 12) return;

    final portal = portals[index];
    switch (portal.type) {
      case PortalType.teamB:
        final points = 100 * comboMultiplier;
        score += points;
        comboMultiplier++;
        final name = portal.familyProfile?.name ?? "Opponent";
        sparkles.add(
          WhackSparkle(
            index: index,
            offset: Offset(_random.nextDouble() * 40 - 20, -50.0),
            text: "+$points",
            isPenalty: false,
          ),
        );
        gameMessage = "WHACKED $name! Combo x$comboMultiplier";
        portal.type = PortalType.empty;
        portal.familyProfile = null;
        break;

      case PortalType.teamA:
        const penalty = 200;
        score = (score - penalty).clamp(0, 999999);
        comboMultiplier = 1;
        final name = portal.familyProfile?.name ?? "Teammate";
        sparkles.add(
          WhackSparkle(
            index: index,
            offset: Offset(_random.nextDouble() * 40 - 20, -50.0),
            text: "Don't tap teammate $name! -$penalty",
            isPenalty: true,
          ),
        );
        gameMessage = "OW! That's teammate $name!";
        portal.type = PortalType.empty;
        portal.familyProfile = null;
        break;

      case PortalType.empty:
        comboMultiplier = 1;
        gameMessage = "MISS!";
        break;
    }
  }

  void _endGame() {
    isGameOver = true;
    isPlaying = false;
    gameMessage = "TIME'S UP! Game Over!";
  }
}
